[[ Bug 19066 ]] Improve LCB's execute script command

runrevmark opened the pull request:
This patch adds a 'in object' and 'with arguments' clause to
LCB's 'execute script' command:

  execute script SCRIPT [ in OBJECT ] [ with ARGUMENTS ]

The arguments can be accessed in the script fragment by using
the param(N) function, and the count via the paramCount()
function.

runrevmark committed:
[[ Bug 19066 ]] Improve LCB's execute script command

This patch adds a 'in object' and 'with arguments' clause to
LCB's 'execute script' command:

  execute script SCRIPT [ in OBJECT ] [ with ARGUMENTS ]

The arguments can be accessed in the script fragment by using
the param(N) function, and the count via the paramCount()
function.

peter-b commented on docs/lcb/notes/19066.md:
```diff
@@ -0,0 +1,8 @@
+## Engine Library
+
+* The `execute script` command has been improved and now allows
+  specification of a target object and list of arguments.
```
It would be great to have an example in the release note!

peter-b commented on engine/src/module-engine.cpp:
```diff
@@ -559,34 +559,83 @@ extern "C" MC_DLLEXPORT_DEF void MCEngineEvalMessageWasNotHandled(bool& r_not_ha
+		}
+		
+		p_object = t_stack->getcurcard();
+	}
```
```c++
MCStack *t_stack = MCdefaultstackptr;
```
is equivalent to
```c++
MCStack *t_stack = MCdefaultstackptr.Get();
```
see [implementation](https://github.com/livecode/livecode/blob/develop/engine/src/object.h#L362).  This can assert.  To avoid a possible assertion failure, I think what you're looking for here is something like:
```c++
if (p_object == nil)
{
    if (MCdefaultstackptr) {
        p_object = MCdefaultstackptr->getcurcard();
    } else {
        MCErrorCreateAndThrow(kMCGenericErrorTypeInfo,
                              "reason", MCSTR("no default stack"), nil);
        return nullptr;
    }
}
```

peter-b commented on engine/src/module-engine.cpp:
```diff
@@ -559,34 +559,83 @@ extern "C" MC_DLLEXPORT_DEF void MCEngineEvalMessageWasNotHandled(bool& r_not_ha
+	}
+	
+	MCExecContext ctxt(p_object, nil, nil);
+	MCParameter *t_params = nil;
```
I suggest:

```c++
MCAutoCustomPointer<MCParameter, MCEngineFreeScriptParameters> t_params;
```

This allows removal of the explicit cleanup and the `goto`.

peter-b commented on tests/lcs/core/engine/_extension.lcb:
```diff
@@ -33,4 +33,19 @@ public handler TestExtensionSupportModule_Handler()
 end handler
 
+public handler TestExtensionExecuteScript_Do(in pScript as String, in pTarget as String, in pParams as List)
+	if pParams is the empty list and pTarget is the empty string then
```
If I remember correctly, this can just be:

```
if pParams is empty and pTarget is empty then
   -- statements
end if
```

Yay for generics!  (Don't change this, it's just an observation)

peter-b reviewed (changes requested):
Please tweak the `MCdefaultstackptr` access!

runrevmark commented:
@peter-b: That should be your suggestions implemented.

peter-b commented on engine/src/engine.lcb:
```diff
@@ -360,17 +365,34 @@ Example:
+Executes the given piece of LiveCode script in the context of the target script
+object, or the current script object if none is specified. The list of arguments
+is accessible from the script fragment using the paramCount() and param()
+functions.
```
Is it possible to explain more clearly how "the current script object" is determined? Is it the same as `my script object`? If so, maybe a cross-reference might make sense

Could you please put `paramCount()` and `param()` in inline code markup?

runrevmark committed:
[[ Bug 19066 ]] Tweak description of 'execute script'

This patch corrects the description of 'execute script' and adds
a note about an anomaly report currently affecting it.

peter-b commented:
@livecode-vulcan review ok 86ac485

livecode-vulcan commented:
💙 review by @peter-b ok 86ac485865bc5b3a519d5563d4d70ceb8f407336

livecode-vulcan commented:
😞 test failure 86ac485865bc5b3a519d5563d4d70ceb8f407336

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/907)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/913)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/905)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/906)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/913)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/904)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/904)

runrevmark committed:
Merge branch 'develop' into bugfix-19066

runrevmark commented:
Self-review after rebase merge.

@livecode-vulcan review ok e0b222d

livecode-vulcan commented:
💙 review by @runrevmark ok e0b222d98aa000330df0bdf461cded00808678dd

livecode-vulcan commented:
😞 test failure e0b222d98aa000330df0bdf461cded00808678dd

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/979)
* try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/985)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/977)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/978)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/985)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/999)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/976)

runrevmark committed:
Merge branch 'develop' into bugfix-19066

runrevmark committed:
[[ Bug 19066 ]] Fix incorrect function call

This patch corrects the call to MCEngineEnsureScriptObjectAccessIsAllowed
which was incorrect after merging a previous incarnation of develop
into the branch.

runrevmark commented:
Self-review after rebase merge.

@livecode-vulcan review ok e0b222d

runrevmark commented:
Self-review after rebase merge.

@livecode-vulcan review ok 653cce5

livecode-vulcan commented:
💙 review by @runrevmark ok 653cce5b9ed3fdc5f68f2a2f32c0fec4b75a7962

livecode-vulcan commented:
😞 test failure 653cce5b9ed3fdc5f68f2a2f32c0fec4b75a7962

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1014)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1020)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1012)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1013)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1020)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1034)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1011)

peter-b committed:
Tweak MCAutoCustomPointer usage to enable MSVC2010 compile

`MCAutoCustomPointer<ElementType,Deleter>` takes a function pointer as
the `Deleter` template argument.  In C++11, this is supposed to work
even if the function lacks external linkage, but this isn't supported
by MSVC2010.  This patch changes `MCEngineFreeScriptParameters()` to
have external linkage in order to fix the MSVC2010 build.

livecodeali commented:
@livecode-vulcan review ok 5931790

livecode-vulcan commented:
💙 review by @livecodeali ok 59317908186edcde23f9d768d35f77d9fc8119c9

livecodeali commented:
@livecode-vulcan retry 5931790

livecode-vulcan commented:
😎 test success 59317908186edcde23f9d768d35f77d9fc8119c9

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1022)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1028)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1020)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1021)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1028)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1042)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1019)
